Filter out bad words

stroncium opened the pull request:
Add bad word filters, fixes #2

Surprisingly, after looking through https://npms.io/search?q=profanity I found out most of it traces back to just 3 or 4 sources, and most of it extremely bad, especially in terms of false positives.

After all, I used `leo-profanity` and `badwords-list`(the package that uses it runs so slow it didn't have enough patience to find how long it will take to process our word list).

I applied some custom magic on top of `badwords-list` to increase hit rate while reducing false positives to reasonable levels.

Commit:
added bad word filters, fixes #2

Commit:
linter

sindresorhus commented on fetch-words.js:
```diff
@@ -4,9 +4,53 @@ const got = require('got');
+leoProfanity.loadDictionary('en');
+
+const badWordsSet = new Set(badWords);
+const badWordsExclusions = new Set(['ass', 'asses', 'anal', 'cum', 'tit', 'bum', 'hore', 'sex', 'cipa', 'butt', 'cock', 'homo', 'eros', 'cox', 'semen', 'damn', 'jap', 'lust', 'hell', 'pron', 'crap']);
```
Can you put each item on a newline?

stroncium replied:
done

sindresorhus commented on fetch-words.js:
```diff
@@ -4,9 +4,53 @@ const got = require('got');
 const url = 'https://raw.githubusercontent.com/atebits/Words/master/Words/en.txt';
 
+const badWords = require('badwords-list').array;
+const leoProfanity = require('leo-profanity');
```
Put `require` statements at the top.

stroncium replied:
done

sindresorhus commented on fetch-words.js:
```diff
@@ -4,9 +4,53 @@ const got = require('got');
 
 const url = 'https://raw.githubusercontent.com/atebits/Words/master/Words/en.txt';
 
+const badWords = require('badwords-list').array;
```
```suggestion
const {array: badWords} = require('badwords-list');
```

stroncium replied:
done

sindresorhus commented on fetch-words.js:
```diff
@@ -4,9 +4,53 @@ const got = require('got');
+		name: 'badwords-list with exclusions substring',
+		fn: word => {
+			for (const badWord of badWordsWithExclusions) {
+				if (word.indexOf(badWord) !== -1) {
```
Use `.includes()` instead. Same in the tests.

stroncium replied:
Done, but I think you're really rushing that `includes` stuff. Not that it was really important, but using `includes` instead of `indexOf` increases run time of whole build by 15% in this case. (`indexOf` is core function and is well optimized everywhere, unlike `includes`, which, as far as I know, is slow in every major implementation.)

sindresorhus replied:
Doesn't matter in this case. The script is only run maybe once a year.

Commit:
words on newlines

Commit:
requires at top

Commit:
style

Commit:
includes instead of indexOf

sindresorhus commented:
I looked through the diff of what's being removed and there seems to be way too many false-positives. For example, all these were removed:

aerospace
backspace
debugger
meatballs
workspace
woodpecker
wristwatch

stroncium commented:
@sindresorhus It's just 0.82% total positives(some of which are false). What rate you find acceptable?
(Original rate was 2.92%, so I reduced false positives rate already by 2.1%, but it takes quite a long time to google all the words to find out if they can be used as profanity.)

sindresorhus commented:
But I don't understand why words like `aerospace` and `workspace` are matching the filter? Seems like there's some large flaw in the matching algorithm.

Commit:
added fetch-words verbose option, ignored spac

stroncium commented:
@sindresorhus because they include `spac`.
Removed `spac` substring test, added `fetch-words` `-v` option which shows why words were excluded.

However the point was that we have almost 275k words, we only exclude around 2k of them. Does it really matter if some of them are excluded as collateral damage? Also at this point it should be possible to just go through all the 2k words we exclude and check them manually for someone who is really fluent in bad words(I am not, sorry).

sindresorhus commented:
I guess my question is, why are you even matching substrings? Can’t we only match the full words? Seems like that would reduce the false-positives to 0.

Commit:
custom bad words list, removed substring matching

stroncium commented:
Because there is no such premade list.
Anyhow, added custom list inside, filters 284 out of 274926 words now(no false positives if we trust `leo-profanity` and `badwords-list`).

sindresorhus commented:
One more thing you could do to catch some more words is to add a `s` to the detected bad word and see if it matches a word. I see a lot of singular words with plural versions.

Commit:
filter tittie

stroncium commented:
It's the same as matching substrings though, we get false positives. (And in fact almost only false positives, there was only a single non-false-positive word it filtered.).

So, I just added this word to the exception list instead.
